presto-raptorx/src/main/java/com/facebook/presto/raptorx/RaptorPageSink.java

```diff
@@ -0,0 +1,333 @@
+        {
+            Block temporalBlock = null;
+            if (temporalColumnIndex.isPresent()) {
+                temporalBlock = page.getBlock(temporalColumnIndex.getAsInt());
+                validateTemporalBlock(temporalBlock);
+            }
+
+            Page bucketArgs = getBucketArgsPage(page);
+            validateBucketingBlock(bucketArgs);
+
+            for (int position = 0; position < page.getPositionCount(); position++) {
+                int bucket = bucketFunction.getBucket(bucketArgs, position);
+                int day = temporalFunction.isPresent() ? temporalFunction.get().getDay(temporalBlock, position) : 0;
+
+                long partition = (((long) bucket) << 32) | toUnsignedLong(day);
+                PageStore store = pageStores.get(partition);
+                if (store == null) {
+                    PageBuffer buffer = createPageBuffer(bucket);
+                    store = new PageStore(buffer, columnTypes);
+                    pageStores.put(partition, store);
+                }
+
+                store.appendPosition(page, position);
+            }
+
+            flushIfNecessary();
+        }
+
+        private Page getBucketArgsPage(Page page)
+        {
+            Block[] blocks = new Block[bucketFields.length];
+            for (int i = 0; i < bucketFields.length; i++) {
+                blocks[i] = page.getBlock(bucketFields[i]);
+            }
+            return new Page(page.getPositionCount(), blocks);
+        }
+
+        public List<PageBuffer> getPageBuffers()
+        {
+            ImmutableList.Builder<PageBuffer> list = ImmutableList.builder();
+            for (PageStore store : pageStores.values()) {
+                store.flushToPageBuffer();
+                store.getPageBuffer().flush();
+                list.add(store.getPageBuffer());
+            }
+            return list.build();
+        }
+
+        public void rollback()
+        {
+            RuntimeException error = new RuntimeException("Exception during rollback");
+            for (PageStore store : pageStores.values()) {
+                try {
+                    store.getPageBuffer().rollback();
+                }
+                catch (Throwable t) {
+                    error.addSuppressed(t);
+                }
+            }
+            if (error.getSuppressed().length > 0) {
+                throw error;
+            }
+        }
+
+        private void flushIfNecessary()
+        {
+            long totalBytes = 0;
+            long maxBytes = 0;
+            PageBuffer maxBuffer = null;
+
+            for (PageStore store : pageStores.values()) {
+                long bytes = store.getUsedMemoryBytes();
+                totalBytes += bytes;
+
+                if ((maxBuffer == null) || (bytes > maxBytes)) {
+                    maxBuffer = store.getPageBuffer();
+                    maxBytes = bytes;
+                }
+            }
+
+            if ((totalBytes > maxBufferBytes) && (maxBuffer != null)) {
```

Comment:
This is tricky, and it's the same as Raptor v1, so I'm going to leave it alone for now.

As long as the added page is less than the buffer size, we will stay balanced, since we can flush a buffer for each added page.

I'm hesitant to add a loop, since the current code is O(n) to find the largest buffer to flush. We could sort the buffers to allow efficiently flushing multiple, but that's substantially more expensive in the common case that only one is flushed.